In [238]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn import preprocessing, linear_model
from sklearn.preprocessing import StandardScaler
import lightgbm as lgb
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.model_selection import KFold
import math

# 各ファイルの確認

In [4]:
%cd C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Description Data

C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Description Data


In [6]:
train_desc_df = pd.read_csv("train_desc_df.csv")
train_desc_df.head()

comp_id    desc_1    desc_2    desc_3    desc_4    desc_5    desc_6  \
0        1 -0.009555  0.002479  0.002927  0.015774 -0.008177 -0.016036   
1        2 -0.009555  0.002479  0.002927  0.015774 -0.008177 -0.016036   
2        3 -0.356706  0.213562  0.252663  0.090735  0.328961 -0.482705   
3        4 -0.294013  0.165262  0.257102  0.421037  0.463214 -0.769155   
4        5 -0.028657  0.157017  0.282709 -2.674227 -0.711383  2.259387   

     desc_7    desc_8    desc_9  ...   desc_41   desc_42   desc_43   desc_44  \
0  0.026697 -0.000106  0.025788  ... -0.029430 -0.029423  0.052256  0.007814   
1  0.026697 -0.000106  0.025788  ... -0.029430 -0.029423  0.052256  0.007814   
2  0.067300  0.384217 -0.147253  ... -0.626436 -0.102908  0.576792 -0.805986   
3  0.159450  0.236385 -0.183974  ... -0.435836  0.052975  0.108777 -0.599593   
4 -0.162175  0.605468  0.712229  ... -1.714496  0.297421 -0.097744  0.000669   

    desc_45   desc_46   desc_47   desc_48   desc_49   desc_50  
0  0.023472 -0.021134  0.003324 -0.004393  0.004294 -0.002223  
1  0.023472 -0.021134  0.003324 -0.004393  0.004294 -0.002223  
2  0.310637  0.422263  0.403496  0.264240  7.057590 -0.252417  
3  0.408430  0.591615  0.415667  0.334706  7.025648 -0.309093  
4 -1.639307  0.244313  0.099975  0.371806 -2.259024 -0.131085  

[5 rows x 51 columns]

In [7]:
test_desc_df = pd.read_csv("public_desc_df.csv")
test_desc_df.head()

comp_id    desc_1    desc_2    desc_3    desc_4    desc_5    desc_6  \
0     3001 -0.266289  0.151930  0.188684  0.935617  0.021999  0.100940   
1     3002 -0.266289  0.151930  0.188684  0.935617  0.021999  0.100940   
2     3003 -0.266289  0.151930  0.188684  0.935617  0.021999  0.100940   
3     3004 -0.298706 -0.022012  0.552924  0.046070  0.044543  0.132187   
4     3005 -0.266289  0.151930  0.188684  0.935617  0.021999  0.100940   

     desc_7    desc_8    desc_9  ...   desc_41   desc_42   desc_43   desc_44  \
0  0.107302 -0.055652  0.023322  ... -1.111625  0.252028 -0.193523  0.346493   
1  0.107302 -0.055652  0.023322  ... -1.111625  0.252028 -0.193523  0.346493   
2  0.107302 -0.055652  0.023322  ... -1.111625  0.252028 -0.193523  0.346493   
3 -0.549770  1.437774  1.286943  ...  0.246156 -1.001942 -1.502183  1.105716   
4  0.107302 -0.055652  0.023322  ... -1.111625  0.252028 -0.193523  0.346493   

    desc_45   desc_46   desc_47   desc_48   desc_49   desc_50  
0 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
1 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
2 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
3 -0.170937  0.181976  0.437707  0.058352 -2.209322  0.104916  
4 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  

[5 rows x 51 columns]

In [8]:
%cd C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Image Data

C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Image Data


In [9]:
train_image_df = pd.read_csv("train_image_df.csv")
train_image_df.head()

comp_id   image_1   image_2   image_3   image_4   image_5   image_6  \
0        1  0.484456  0.036087 -0.149328  0.144475 -0.617386  0.035018   
1        2  0.074533 -0.015655 -0.016286 -0.480964  0.687917  0.037131   
2        3 -0.396809  0.021490 -1.723037  0.666147 -0.631924  0.047724   
3        4  0.995316  0.012766  0.387472 -0.684791 -0.209261  0.013654   
4        5 -0.611648  0.000860 -0.572393  0.894287 -0.191228 -0.111583   

    image_7   image_8   image_9  ...  image_3991  image_3992  image_3993  \
0  0.169174 -0.005733 -0.104908  ...   -0.090408   -0.032915   -0.010857   
1 -0.149725 -0.002098  0.099383  ...   -0.625816   -0.009616    0.004558   
2  0.336041 -0.003904  0.039683  ...   -0.078059    0.086320   -0.005606   
3  0.138517  0.005586  0.079277  ...   -0.190714   -0.005452   -0.003889   
4 -0.011111  0.003231 -0.160582  ...   -0.040437   -0.142134   -0.022771   

   image_3994  image_3995  image_3996  image_3997  image_3998  image_3999  \
0   -0.015954    0.123416    0.195833   -0.066449    0.115818   -0.005140   
1    0.008310   -0.173496    0.133518   -0.488898    0.084832    0.146566   
2    0.002414   -0.164493    0.218473    0.151292   -0.076860    0.008321   
3    0.013430    0.344406    0.038220    0.210446   -0.145128   -0.019172   
4   -0.013134   -0.024869    0.125129    0.215967    0.005288    0.145715   

   image_4000  
0    0.017278  
1    0.007997  
2    0.012555  
3    0.000532  
4    0.001425  

[5 rows x 4001 columns]

In [10]:
test_image_df = pd.read_csv("public_image_df.csv")
test_image_df.head()

comp_id   image_1   image_2   image_3   image_4   image_5   image_6  \
0     3001 -0.288211  0.007964 -0.318695 -0.519448 -0.707238  0.018358   
1     3002 -0.053049 -0.019362 -0.942308  0.971880 -0.145869  0.190100   
2     3003  0.527363  0.027754 -1.644639 -0.362191 -0.275900  0.095907   
3     3004  0.709023  0.002011 -0.968376  0.478994 -0.451566  0.217366   
4     3005 -0.349026 -0.060301 -1.099632  1.475574 -0.076292  0.153835   

    image_7   image_8   image_9  ...  image_3991  image_3992  image_3993  \
0 -0.127491  0.006911 -0.163305  ...    0.095390   -0.005706    0.011662   
1  0.323404  0.001693  0.052303  ...   -0.152575   -0.015299    0.006474   
2 -0.083026 -0.002009  0.041976  ...   -0.139857    0.056012    0.010546   
3 -0.036373  0.004914  0.133615  ...   -0.343643    0.072151    0.045109   
4 -0.143689  0.006389  0.058670  ...    0.158350   -0.024394    0.002807   

   image_3994  image_3995  image_3996  image_3997  image_3998  image_3999  \
0    0.027501    0.202042    0.031049    0.058993    0.085795   -0.033876   
1    0.008779    0.064745   -0.130691    0.519738   -0.009965    0.226757   
2   -0.013047    0.395852    0.064757    0.061725    0.081311    0.094162   
3   -0.009794   -0.002994    0.142427   -0.001836    0.116891    0.078607   
4   -0.004787   -0.012235    0.022360    0.034833    0.084962   -0.092006   

   image_4000  
0   -0.010047  
1    0.001792  
2   -0.015818  
3   -0.000430  
4    0.010375  

[5 rows x 4001 columns]

In [11]:
%cd C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Metadata

C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Metadata


In [12]:
train_meta_df = pd.read_csv("train_meta_df.csv")
train_meta_df.head()

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0        1       False   True  1.77778        86         3     True   
1        2       False   True  1.33333      1129         2     True   
2        3       False   True  1.76667      1163         2     True   
3        4       False   True  1.77778      1326         2     True   
4        5       False   True  1.77273      2612         2     True   

   partner_active  n_likes  views  n_tags  n_formats  dayofweek  hour  
0           False        0    290       3          0          5     6  
1           False        0   1443       0          2          5     6  
2           False        0    329       0          1          5     6  
3           False        0     63       0          3          5     6  
4           False        3     37       0          0          5     6

In [13]:
test_meta_df = pd.read_csv("public_meta_df.csv")
test_meta_df.head()

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0     3001       False   True  1.33333      1675         2    False   
1     3002       False   True  1.33333      1479         2    False   
2     3003       False   True  1.33333      1505         2    False   
3     3004       False   True  1.77778        50         2     True   
4     3005       False   True  1.33333      1543         2    False   

   partner_active  n_likes  n_tags  n_formats  dayofweek  hour  
0           False        0      10          2          2     3  
1           False        0      10          2          2     3  
2           False        0      10          2          2     3  
3            True        1       6          5          2     3  
4           False        0       6          2          2     3

In [14]:
%cd C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Title Data

C:\Users\Takeshi\OneDrive\Bitgrid\Dataset\Title Data


In [15]:
train_title_df = pd.read_csv("train_title_df.csv")
train_title_df.head()

comp_id   title_1   title_2   title_3   title_4   title_5   title_6  \
0        1 -0.977637 -0.543310  0.079403  0.205560 -1.497104  0.230466   
1        2  0.041873  0.644655  0.140869 -0.664714 -0.062992  0.240086   
2        3 -0.905595  0.097928  0.111981 -0.064753 -1.566516  0.140807   
3        4  0.069220  0.411544  0.177700 -0.739998 -0.932620 -0.717982   
4        5  0.503560 -0.210970 -0.085412  0.549240 -0.019521 -0.255597   

    title_7   title_8   title_9  ...  title_41  title_42  title_43  title_44  \
0  0.566112 -0.662264  0.130342  ...  0.059434 -0.272340 -0.030180 -0.109040   
1 -1.017593  1.019706  0.340133  ...  0.242961 -0.404538  0.335224 -0.158068   
2 -0.946588 -0.342442  0.031226  ...  0.414191 -0.457968  0.157185 -0.116985   
3 -0.406487 -0.267096  0.114066  ...  0.211731 -0.812590  0.258014 -0.180143   
4  0.841144  0.250485  0.224235  ... -0.517912  0.543584  0.492052 -0.204867   

   title_45  title_46  title_47  title_48  title_49  title_50  
0 -0.033425  0.728739  0.142030  0.397541  1.270722 -0.816523  
1 -0.178225  0.351125  0.959121  0.673848 -0.530106  0.133466  
2  0.404937 -0.220503  0.443171  0.643445  0.357957  0.525154  
3 -0.564691 -0.515378  0.613686  0.817547  0.325042  0.254415  
4 -0.011538 -0.328607 -0.150055 -0.992373 -0.266003 -0.037134  

[5 rows x 51 columns]

In [16]:
test_title_df = pd.read_csv("public_title_df.csv")
test_title_df.head()

comp_id   title_1   title_2   title_3   title_4   title_5   title_6  \
0     3001  0.363698  0.309343  0.225138  0.318105 -0.523210  0.143360   
1     3002  0.218130  1.117077  0.122527 -0.005565 -0.158261  0.390527   
2     3003  0.209271  0.785080  0.287104  0.508236  0.306759  0.505576   
3     3004  1.160773 -0.491947  0.188722 -0.100140 -1.738414 -0.105218   
4     3005  1.338603  0.599431  0.519767 -0.198956  0.850357  0.039661   

    title_7   title_8   title_9  ...  title_41  title_42  title_43  title_44  \
0 -0.012829  0.407408 -0.048370  ...  0.532943  0.194463 -0.635401 -0.419310   
1  0.189782  0.685502 -0.383055  ...  0.548985  0.066296 -0.525967 -0.674784   
2 -0.172887  0.324774 -0.032752  ...  0.803655  0.199135 -0.580919 -0.569989   
3  0.400427 -0.216526 -0.371494  ...  0.125851  0.735062  0.330460  0.646128   
4  0.248881  0.885659  0.245418  ...  0.531473 -0.406249 -0.507858 -0.187037   

   title_45  title_46  title_47  title_48  title_49  title_50  
0  0.142901 -0.523376  1.204893 -1.413345 -0.096672  0.078235  
1  0.151137 -1.065329  0.751027 -1.075753 -0.009242  0.464259  
2  0.033141 -0.658906  0.956579 -0.885134 -0.233097  0.235301  
3  0.809633  0.162228  0.274926 -0.085908  0.217528  1.031500  
4  0.802942 -0.472328  0.897817 -1.227795  0.746802 -0.379223  

[5 rows x 51 columns]

# Metadetaの内訳確認等

In [17]:
# Null値

train_meta_df.isnull().sum()

comp_id           0
ad_blocked        0
embed             0
ratio             0
duration          0
language          0
partner           0
partner_active    0
n_likes           0
views             0
n_tags            0
n_formats         0
dayofweek         0
hour              0
dtype: int64

In [18]:
# 情報の内容確認

train_meta_df.describe()

comp_id        ratio      duration     language      n_likes  \
count  3000.000000  3000.000000   3000.000000  3000.000000  3000.000000   
mean   1500.500000     1.650824    545.717667     2.942000     0.186333   
std     866.169729     0.307880   1003.643982     1.898904     0.986883   
min       1.000000     0.550000      0.000000     1.000000     0.000000   
25%     750.750000     1.739130     61.000000     2.000000     0.000000   
50%    1500.500000     1.777780    143.000000     2.000000     0.000000   
75%    2250.250000     1.777780    421.250000     4.000000     0.000000   
max    3000.000000     3.200000  20068.000000    10.000000    20.000000   

             views       n_tags    n_formats    dayofweek         hour  
count  3000.000000  3000.000000  3000.000000  3000.000000  3000.000000  
mean    548.708333     3.826000     4.011667     3.933667     8.143333  
std    1212.155134     5.169329     1.189398     2.175663     6.825515  
min      20.000000     0.000000     0.000000     1.000000     0.000000  
25%      58.000000     0.000000     3.000000     2.000000     4.000000  
50%     128.000000     1.000000     4.000000     3.000000     5.000000  
75%     348.000000     8.000000     5.000000     7.000000    13.000000  
max    9545.000000    60.000000     8.000000     7.000000    23.000000

In [19]:
# hourの内訳確認

train_meta_df["hour"].unique()

array([ 6,  5,  4,  3,  2,  1,  0, 23,  7, 22, 21, 20, 19, 17, 16, 15, 14,
       13, 12, 11, 10,  9,  8, 18], dtype=int64)

In [20]:
# partnerの内訳確認

train_meta_df["partner"].value_counts()

True     1806
False    1194
Name: partner, dtype: int64

割とcertifiedされていないvideoがある事が分かる。

In [21]:
# partner_activeの内訳確認

train_meta_df["partner_active"].value_counts()

False    2317
True      683
Name: partner_active, dtype: int64

partnerがアクティブで無いものも一定数ある事が分かる。

In [22]:
# n_formatsの内訳

train_meta_df["n_formats"].value_counts()

4    1039
5     768
3     668
6     271
2     142
1      88
0      19
7       3
8       2
Name: n_formats, dtype: int64

formatには8つの種類がある。

In [43]:
# ratioの内訳

df_train["ratio"].value_counts()

1.777780    1549
1.333330     413
1.766670     299
1.785710      90
1.764710      73
            ... 
1.781610       1
1.774510       1
1.466670       1
2.370370       1
1.333333       1
Name: ratio, Length: 128, dtype: int64

細かく128種類ある。

# Bool値を数値化する

In [23]:
train_meta_df[["ad_blocked","embed", "partner", "partner_active"]] = train_meta_df[["ad_blocked","embed", "partner", "partner_active"]].astype(np.int)
train_meta_df

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0           1           0      1  1.77778        86         3        1   
1           2           0      1  1.33333      1129         2        1   
2           3           0      1  1.76667      1163         2        1   
3           4           0      1  1.77778      1326         2        1   
4           5           0      1  1.77273      2612         2        1   
...       ...         ...    ...      ...       ...       ...      ...   
2995     2996           0      1  1.76667       138         2        0   
2996     2997           0      1  1.77778       113         2        0   
2997     2998           0      1  1.77778        30         2        0   
2998     2999           0      1  1.77778       115         2        0   
2999     3000           0      1  1.33333       644         4        0   

      partner_active  n_likes  views  n_tags  n_formats  dayofweek  hour  
0                  0        0    290       3          0          5     6  
1                  0        0   1443       0          2          5     6  
2                  0        0    329       0          1          5     6  
3                  0        0     63       0          3          5     6  
4                  0        3     37       0          0          5     6  
...              ...      ...    ...     ...        ...        ...   ...  
2995               0        0     20       8          3          2     4  
2996               0        0    234       1          4          2     4  
2997               0        0     97       6          4          2     4  
2998               0        0     77       5          4          2     3  
2999               0        0     59       1          4          2     3  

[3000 rows x 14 columns]

In [24]:
test_meta_df[["ad_blocked","embed", "partner", "partner_active"]] = test_meta_df[["ad_blocked","embed", "partner", "partner_active"]].astype(np.int)
test_meta_df

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0       3001           0      1  1.33333      1675         2        0   
1       3002           0      1  1.33333      1479         2        0   
2       3003           0      1  1.33333      1505         2        0   
3       3004           0      1  1.77778        50         2        1   
4       3005           0      1  1.33333      1543         2        0   
..       ...         ...    ...      ...       ...       ...      ...   
981     3982           0      1  1.77778        46         3        1   
982     3983           0      1  1.33333        91         2        0   
983     3984           0      1  1.77778        72         3        1   
984     3985           0      1  1.33333        82         7        0   
985     3986           0      1  1.34091        88         2        0   

     partner_active  n_likes  n_tags  n_formats  dayofweek  hour  
0                 0        0      10          2          2     3  
1                 0        0      10          2          2     3  
2                 0        0      10          2          2     3  
3                 1        1       6          5          2     3  
4                 0        0       6          2          2     3  
..              ...      ...     ...        ...        ...   ...  
981               1        0       2          4          6    13  
982               0        0       1          2          6    13  
983               1        0       2          4          6    13  
984               0        0       9          3          6    13  
985               0        0       4          3          6    13  

[986 rows x 13 columns]

# Meta data, Image data, Title data, Description dataを結合させる

In [25]:
# Meta dataとImage Dataの結合(train)

df_train = pd.merge(train_meta_df, train_image_df, on = "comp_id")
df_train

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0           1           0      1  1.77778        86         3        1   
1           2           0      1  1.33333      1129         2        1   
2           3           0      1  1.76667      1163         2        1   
3           4           0      1  1.77778      1326         2        1   
4           5           0      1  1.77273      2612         2        1   
...       ...         ...    ...      ...       ...       ...      ...   
2995     2996           0      1  1.76667       138         2        0   
2996     2997           0      1  1.77778       113         2        0   
2997     2998           0      1  1.77778        30         2        0   
2998     2999           0      1  1.77778       115         2        0   
2999     3000           0      1  1.33333       644         4        0   

      partner_active  n_likes  views  ...  image_3991  image_3992  image_3993  \
0                  0        0    290  ...   -0.090408   -0.032915   -0.010857   
1                  0        0   1443  ...   -0.625816   -0.009616    0.004558   
2                  0        0    329  ...   -0.078059    0.086320   -0.005606   
3                  0        0     63  ...   -0.190714   -0.005452   -0.003889   
4                  0        3     37  ...   -0.040437   -0.142134   -0.022771   
...              ...      ...    ...  ...         ...         ...         ...   
2995               0        0     20  ...   -0.489864    0.015473   -0.036980   
2996               0        0    234  ...    0.225417   -0.137557    0.011404   
2997               0        0     97  ...   -0.171302   -0.044187    0.000784   
2998               0        0     77  ...    0.031655   -0.040731   -0.019667   
2999               0        0     59  ...    0.167214    0.230765    0.014820   

      image_3994  image_3995  image_3996  image_3997  image_3998  image_3999  \
0      -0.015954    0.123416    0.195833   -0.066449    0.115818   -0.005140   
1       0.008310   -0.173496    0.133518   -0.488898    0.084832    0.146566   
2       0.002414   -0.164493    0.218473    0.151292   -0.076860    0.008321   
3       0.013430    0.344406    0.038220    0.210446   -0.145128   -0.019172   
4      -0.013134   -0.024869    0.125129    0.215967    0.005288    0.145715   
...          ...         ...         ...         ...         ...         ...   
2995    0.009553    0.300495   -0.036730    0.020080   -0.022211    0.078571   
2996   -0.020122   -0.381863   -0.244808    0.076713    0.062716    0.016840   
2997   -0.005235    0.524747   -0.132455   -0.102490    0.003506    0.106999   
2998   -0.015320    0.087113    0.107709    0.219753   -0.065209    0.211636   
2999    0.014386   -0.220913    0.155001    0.149516   -0.118682    0.024679   

      image_4000  
0       0.017278  
1       0.007997  
2       0.012555  
3       0.000532  
4       0.001425  
...          ...  
2995   -0.046191  
2996   -0.002579  
2997    0.001011  
2998   -0.001608  
2999   -0.009011  

[3000 rows x 4014 columns]

In [26]:
# Meta dataとImage Dataの結合(test)

df_test = pd.merge(test_meta_df, test_image_df, on = "comp_id")
df_test

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0       3001           0      1  1.33333      1675         2        0   
1       3002           0      1  1.33333      1479         2        0   
2       3003           0      1  1.33333      1505         2        0   
3       3004           0      1  1.77778        50         2        1   
4       3005           0      1  1.33333      1543         2        0   
..       ...         ...    ...      ...       ...       ...      ...   
981     3982           0      1  1.77778        46         3        1   
982     3983           0      1  1.33333        91         2        0   
983     3984           0      1  1.77778        72         3        1   
984     3985           0      1  1.33333        82         7        0   
985     3986           0      1  1.34091        88         2        0   

     partner_active  n_likes  n_tags  ...  image_3991  image_3992  image_3993  \
0                 0        0      10  ...    0.095390   -0.005706    0.011662   
1                 0        0      10  ...   -0.152575   -0.015299    0.006474   
2                 0        0      10  ...   -0.139857    0.056012    0.010546   
3                 1        1       6  ...   -0.343643    0.072151    0.045109   
4                 0        0       6  ...    0.158350   -0.024394    0.002807   
..              ...      ...     ...  ...         ...         ...         ...   
981               1        0       2  ...    0.074584   -0.116341   -0.033850   
982               0        0       1  ...   -0.490430    0.065782   -0.013688   
983               1        0       2  ...   -0.206422   -0.017027   -0.020116   
984               0        0       9  ...   -0.298586    0.029438    0.007002   
985               0        0       4  ...   -0.086795   -0.031632   -0.026254   

     image_3994  image_3995  image_3996  image_3997  image_3998  image_3999  \
0      0.027501    0.202042    0.031049    0.058993    0.085795   -0.033876   
1      0.008779    0.064745   -0.130691    0.519738   -0.009965    0.226757   
2     -0.013047    0.395852    0.064757    0.061725    0.081311    0.094162   
3     -0.009794   -0.002994    0.142427   -0.001836    0.116891    0.078607   
4     -0.004787   -0.012235    0.022360    0.034833    0.084962   -0.092006   
..          ...         ...         ...         ...         ...         ...   
981   -0.081973   -0.132744   -0.071409   -0.060938   -0.025997    0.007047   
982   -0.008404   -0.070429   -0.020449    0.087405   -0.017570   -0.100766   
983    0.025695   -0.064006    0.032865   -0.106347    0.051806    0.014432   
984    0.009091   -0.221092    0.002472   -0.133447    0.093916    0.004973   
985   -0.012215    0.179908   -0.075768   -0.131879    0.050861    0.119456   

     image_4000  
0     -0.010047  
1      0.001792  
2     -0.015818  
3     -0.000430  
4      0.010375  
..          ...  
981    0.011844  
982    0.005535  
983    0.037174  
984   -0.002524  
985   -0.007984  

[986 rows x 4013 columns]

In [27]:
# dfとTitle Dataの結合(train)

df_train = pd.merge(df_train, train_title_df, on = "comp_id")
df_train

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0           1           0      1  1.77778        86         3        1   
1           2           0      1  1.33333      1129         2        1   
2           3           0      1  1.76667      1163         2        1   
3           4           0      1  1.77778      1326         2        1   
4           5           0      1  1.77273      2612         2        1   
...       ...         ...    ...      ...       ...       ...      ...   
2995     2996           0      1  1.76667       138         2        0   
2996     2997           0      1  1.77778       113         2        0   
2997     2998           0      1  1.77778        30         2        0   
2998     2999           0      1  1.77778       115         2        0   
2999     3000           0      1  1.33333       644         4        0   

      partner_active  n_likes  views  ...  title_41  title_42  title_43  \
0                  0        0    290  ...  0.059434 -0.272340 -0.030180   
1                  0        0   1443  ...  0.242961 -0.404538  0.335224   
2                  0        0    329  ...  0.414191 -0.457968  0.157185   
3                  0        0     63  ...  0.211731 -0.812590  0.258014   
4                  0        3     37  ... -0.517912  0.543584  0.492052   
...              ...      ...    ...  ...       ...       ...       ...   
2995               0        0     20  ...  0.334074 -0.188084 -0.673507   
2996               0        0    234  ...  0.380635  0.705248  0.664459   
2997               0        0     97  ...  0.578262 -0.616080  0.085226   
2998               0        0     77  ... -0.893163 -1.703396  0.009605   
2999               0        0     59  ...  0.423330 -0.971956  0.225077   

      title_44  title_45  title_46  title_47  title_48  title_49  title_50  
0    -0.109040 -0.033425  0.728739  0.142030  0.397541  1.270722 -0.816523  
1    -0.158068 -0.178225  0.351125  0.959121  0.673848 -0.530106  0.133466  
2    -0.116985  0.404937 -0.220503  0.443171  0.643445  0.357957  0.525154  
3    -0.180143 -0.564691 -0.515378  0.613686  0.817547  0.325042  0.254415  
4    -0.204867 -0.011538 -0.328607 -0.150055 -0.992373 -0.266003 -0.037134  
...        ...       ...       ...       ...       ...       ...       ...  
2995 -0.320320 -0.119507 -2.042197  0.104381 -1.097715  0.589811 -0.030453  
2996 -0.444807 -0.027819 -1.750902  0.585283 -1.624563  0.267947 -0.320521  
2997 -0.468162 -0.069771  0.042086 -0.056725 -1.810434 -0.489969 -0.228624  
2998 -0.487247 -0.383803  0.818402 -0.648870  3.659653  0.871033  0.720780  
2999  0.134018 -0.227706 -0.395383 -0.570513 -1.255753  0.379903  0.199309  

[3000 rows x 4064 columns]

In [28]:
# dfとTitle Dataの結合(test)

df_test = pd.merge(df_test, test_title_df, on = "comp_id")
df_test

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0       3001           0      1  1.33333      1675         2        0   
1       3002           0      1  1.33333      1479         2        0   
2       3003           0      1  1.33333      1505         2        0   
3       3004           0      1  1.77778        50         2        1   
4       3005           0      1  1.33333      1543         2        0   
..       ...         ...    ...      ...       ...       ...      ...   
981     3982           0      1  1.77778        46         3        1   
982     3983           0      1  1.33333        91         2        0   
983     3984           0      1  1.77778        72         3        1   
984     3985           0      1  1.33333        82         7        0   
985     3986           0      1  1.34091        88         2        0   

     partner_active  n_likes  n_tags  ...  title_41  title_42  title_43  \
0                 0        0      10  ...  0.532943  0.194463 -0.635401   
1                 0        0      10  ...  0.548985  0.066296 -0.525967   
2                 0        0      10  ...  0.803655  0.199135 -0.580919   
3                 1        1       6  ...  0.125851  0.735062  0.330460   
4                 0        0       6  ...  0.531473 -0.406249 -0.507858   
..              ...      ...     ...  ...       ...       ...       ...   
981               1        0       2  ... -0.294802  0.058419  0.605834   
982               0        0       1  ...  0.172808 -1.670855 -0.129262   
983               1        0       2  ... -0.288506  0.624687  0.849840   
984               0        0       9  ...  1.439242 -0.266624  0.561649   
985               0        0       4  ... -0.249079  1.244390  0.325135   

     title_44  title_45  title_46  title_47  title_48  title_49  title_50  
0   -0.419310  0.142901 -0.523376  1.204893 -1.413345 -0.096672  0.078235  
1   -0.674784  0.151137 -1.065329  0.751027 -1.075753 -0.009242  0.464259  
2   -0.569989  0.033141 -0.658906  0.956579 -0.885134 -0.233097  0.235301  
3    0.646128  0.809633  0.162228  0.274926 -0.085908  0.217528  1.031500  
4   -0.187037  0.802942 -0.472328  0.897817 -1.227795  0.746802 -0.379223  
..        ...       ...       ...       ...       ...       ...       ...  
981 -0.465382  0.309643 -0.018230 -0.439516  0.429129 -1.036004  1.198265  
982 -0.187327  0.720298  1.484852  0.000885  1.031016 -0.930964 -0.352015  
983 -0.787343  0.543122 -0.172172 -0.565166  0.508582 -0.855529  0.981812  
984  0.613700  0.890145  1.952276  0.774547  1.772967  0.367391 -0.789109  
985  0.139387 -0.106472  0.013029 -1.054775  1.202521 -0.918951  0.079340  

[986 rows x 4063 columns]

In [29]:
# dfとDescription Dataの結合(train)

df_train = pd.merge(df_train, train_desc_df, on = "comp_id")
df_train

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0           1           0      1  1.77778        86         3        1   
1           2           0      1  1.33333      1129         2        1   
2           3           0      1  1.76667      1163         2        1   
3           4           0      1  1.77778      1326         2        1   
4           5           0      1  1.77273      2612         2        1   
...       ...         ...    ...      ...       ...       ...      ...   
2995     2996           0      1  1.76667       138         2        0   
2996     2997           0      1  1.77778       113         2        0   
2997     2998           0      1  1.77778        30         2        0   
2998     2999           0      1  1.77778       115         2        0   
2999     3000           0      1  1.33333       644         4        0   

      partner_active  n_likes  views  ...   desc_41   desc_42   desc_43  \
0                  0        0    290  ... -0.029430 -0.029423  0.052256   
1                  0        0   1443  ... -0.029430 -0.029423  0.052256   
2                  0        0    329  ... -0.626436 -0.102908  0.576792   
3                  0        0     63  ... -0.435836  0.052975  0.108777   
4                  0        3     37  ... -1.714496  0.297421 -0.097744   
...              ...      ...    ...  ...       ...       ...       ...   
2995               0        0     20  ...  0.190193 -0.755953 -0.231644   
2996               0        0    234  ... -1.433418  1.234688  0.406196   
2997               0        0     97  ... -0.115696 -0.746977  0.685888   
2998               0        0     77  ... -1.174133 -1.123509  0.387533   
2999               0        0     59  ...  0.961226  0.267603  0.389177   

       desc_44   desc_45   desc_46   desc_47   desc_48   desc_49   desc_50  
0     0.007814  0.023472 -0.021134  0.003324 -0.004393  0.004294 -0.002223  
1     0.007814  0.023472 -0.021134  0.003324 -0.004393  0.004294 -0.002223  
2    -0.805986  0.310637  0.422263  0.403496  0.264240  7.057590 -0.252417  
3    -0.599593  0.408430  0.591615  0.415667  0.334706  7.025648 -0.309093  
4     0.000669 -1.639307  0.244313  0.099975  0.371806 -2.259024 -0.131085  
...        ...       ...       ...       ...       ...       ...       ...  
2995  0.308153 -0.373529  0.876928  0.422415  0.074538 -3.646102 -0.302841  
2996  0.877129 -1.218880  1.145661 -0.112073  0.230396 -3.713165 -0.525256  
2997  0.203295 -0.379776  0.512881 -0.275278  0.459139  2.927374  0.219621  
2998 -0.117955  1.226050 -1.122658 -0.056242 -0.078198 -4.572043 -0.803208  
2999  0.024658  0.593088 -0.912221 -0.960675 -0.498283  0.549810  0.213448  

[3000 rows x 4114 columns]

In [30]:
# dfとDescription Dataの結合(test)

df_test = pd.merge(df_test, test_desc_df, on = "comp_id")
df_test

comp_id  ad_blocked  embed    ratio  duration  language  partner  \
0       3001           0      1  1.33333      1675         2        0   
1       3002           0      1  1.33333      1479         2        0   
2       3003           0      1  1.33333      1505         2        0   
3       3004           0      1  1.77778        50         2        1   
4       3005           0      1  1.33333      1543         2        0   
..       ...         ...    ...      ...       ...       ...      ...   
981     3982           0      1  1.77778        46         3        1   
982     3983           0      1  1.33333        91         2        0   
983     3984           0      1  1.77778        72         3        1   
984     3985           0      1  1.33333        82         7        0   
985     3986           0      1  1.34091        88         2        0   

     partner_active  n_likes  n_tags  ...   desc_41   desc_42   desc_43  \
0                 0        0      10  ... -1.111625  0.252028 -0.193523   
1                 0        0      10  ... -1.111625  0.252028 -0.193523   
2                 0        0      10  ... -1.111625  0.252028 -0.193523   
3                 1        1       6  ...  0.246156 -1.001942 -1.502183   
4                 0        0       6  ... -1.111625  0.252028 -0.193523   
..              ...      ...     ...  ...       ...       ...       ...   
981               1        0       2  ... -0.429883  2.516276  0.551086   
982               0        0       1  ...  2.279406  1.579310  6.187336   
983               1        0       2  ... -0.852350  2.475394  0.530240   
984               0        0       9  ... -1.924031 -0.009321  1.078347   
985               0        0       4  ...  1.163483 -1.815711  0.958069   

      desc_44   desc_45   desc_46   desc_47   desc_48   desc_49   desc_50  
0    0.346493 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
1    0.346493 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
2    0.346493 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
3    1.105716 -0.170937  0.181976  0.437707  0.058352 -2.209322  0.104916  
4    0.346493 -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
..        ...       ...       ...       ...       ...       ...       ...  
981 -0.445840 -1.834950 -1.343484  1.518998  0.690137  2.181449  0.421095  
982 -0.259915  2.230830 -3.986988 -0.375218 -0.185296 -3.666707  0.629068  
983 -0.414277 -1.807584 -1.542420  1.573145  0.537607  2.239486  0.168694  
984  0.409481  1.172790  0.548644 -0.209369 -0.329074  0.300092 -0.437185  
985  1.194779  1.158204 -0.711968 -0.568707 -0.240866  0.391280 -0.164088  

[986 rows x 4113 columns]

# 重回帰分析にかける

In [52]:
# metadataを目的変数(views)と説明変数(それ以外)に分ける

df_exp = df_train.drop(["views","comp_id"], axis = 1)
df_exp

ad_blocked  embed    ratio  duration  language  partner  partner_active  \
0              0      1  1.77778        86         3        1               0   
1              0      1  1.33333      1129         2        1               0   
2              0      1  1.76667      1163         2        1               0   
3              0      1  1.77778      1326         2        1               0   
4              0      1  1.77273      2612         2        1               0   
...          ...    ...      ...       ...       ...      ...             ...   
2995           0      1  1.76667       138         2        0               0   
2996           0      1  1.77778       113         2        0               0   
2997           0      1  1.77778        30         2        0               0   
2998           0      1  1.77778       115         2        0               0   
2999           0      1  1.33333       644         4        0               0   

      n_likes  n_tags  n_formats  ...   desc_41   desc_42   desc_43   desc_44  \
0           0       3          0  ... -0.029430 -0.029423  0.052256  0.007814   
1           0       0          2  ... -0.029430 -0.029423  0.052256  0.007814   
2           0       0          1  ... -0.626436 -0.102908  0.576792 -0.805986   
3           0       0          3  ... -0.435836  0.052975  0.108777 -0.599593   
4           3       0          0  ... -1.714496  0.297421 -0.097744  0.000669   
...       ...     ...        ...  ...       ...       ...       ...       ...   
2995        0       8          3  ...  0.190193 -0.755953 -0.231644  0.308153   
2996        0       1          4  ... -1.433418  1.234688  0.406196  0.877129   
2997        0       6          4  ... -0.115696 -0.746977  0.685888  0.203295   
2998        0       5          4  ... -1.174133 -1.123509  0.387533 -0.117955   
2999        0       1          4  ...  0.961226  0.267603  0.389177  0.024658   

       desc_45   desc_46   desc_47   desc_48   desc_49   desc_50  
0     0.023472 -0.021134  0.003324 -0.004393  0.004294 -0.002223  
1     0.023472 -0.021134  0.003324 -0.004393  0.004294 -0.002223  
2     0.310637  0.422263  0.403496  0.264240  7.057590 -0.252417  
3     0.408430  0.591615  0.415667  0.334706  7.025648 -0.309093  
4    -1.639307  0.244313  0.099975  0.371806 -2.259024 -0.131085  
...        ...       ...       ...       ...       ...       ...  
2995 -0.373529  0.876928  0.422415  0.074538 -3.646102 -0.302841  
2996 -1.218880  1.145661 -0.112073  0.230396 -3.713165 -0.525256  
2997 -0.379776  0.512881 -0.275278  0.459139  2.927374  0.219621  
2998  1.226050 -1.122658 -0.056242 -0.078198 -4.572043 -0.803208  
2999  0.593088 -0.912221 -0.960675 -0.498283  0.549810  0.213448  

[3000 rows x 4112 columns]

In [53]:
# train_test_splitでtrainデータを分ける

X_train, X_test, y_train, y_test = train_test_split(df_exp, train_meta_df["views"])

In [54]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [55]:
model.score(X_train, y_train)

0.9988475203044801

In [56]:
model.score(X_test, y_test)

-8.607219476729197e+16

過学習が起きてしまっている。  
そこで、ElasticNetを用いてみる。

In [80]:
clf_el= linear_model.ElasticNet(alpha=1.0, l1_ratio=0.5)
clf_el.fit(X_train, y_train)

ElasticNet(alpha=1.0, copy_X=True, fit_intercept=True, l1_ratio=0.5,
           max_iter=1000, normalize=False, positive=False, precompute=False,
           random_state=None, selection='cyclic', tol=0.0001, warm_start=False)

In [82]:
clf_el.score(X_train, y_train)

0.615421208432434

In [83]:
clf_el.score(X_test, y_test)

-0.03847343066644893

訓練データの精度は下がったが、テストデータの結果を見るともう少し高い精度を求めたい。  
  
次はRidge回帰を用いてみる。

In [84]:
clf_ri = linear_model.Ridge(alpha=1.0)
clf_ri.fit(X_train, y_train)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [85]:
clf_ri.score(X_train, y_train)

0.9966568897447143

In [86]:
clf_ri.score(X_test, y_test)

-2.6968570615611718

似たような結果になっている。  
  
そこで、次はlight gbmを用いてみる。

# light gbmを用いてみる

In [88]:
lgb_train = lgb.Dataset(X_train, y_train)
lgb_eval = lgb.Dataset(X_test, y_test)

In [92]:
params = {'metric': 'rmse',
          'max_depth' : 8}

In [93]:
gbm = lgb.train(params,
                lgb_train,
                valid_sets=lgb_eval,
                num_boost_round=1000,
                early_stopping_rounds=100,
                verbose_eval=50)

Training until validation scores don't improve for 100 rounds
[50]	valid_0's rmse: 1151.46
[100]	valid_0's rmse: 1166.43
Early stopping, best iteration is:
[12]	valid_0's rmse: 1124.34


R2係数とRMSEを出してみる。

In [180]:
# X_testに対しての予測値
predicted_X_train = gbm.predict(X_train)
predicted_X_test = gbm.predict(X_test)

# それを元にしたR2係数とRMSE
R2_X_train = r2_score(y_train, predicted_X_train)
R2_X_test = r2_score(y_test, predicted_X_test)
RMSE_X_train = np.sqrt(mean_squared_error(y_train, predicted_X_train))
RMSE_X_test = np.sqrt(mean_squared_error(y_test, predicted_X_test))

print("訓練データR2係数 :" + str(R2_X_train))
print("テストデータR2係数 :" + str(R2_X_test))
print("訓練データRMSE :" + str(RMSE_X_train))
print("テストデータRMSE :" + str(RMSE_X_test))

訓練データR2係数 :0.4724680786077934
テストデータR2係数 :0.11764133018430589
訓練データRMSE :883.859267835924
テストデータRMSE :1124.3423572341815


一応結果としては重回帰分析よりは現実的な数字になった。

# 交差検証でモデルを構築する

より精度を高めるため、交差検証でモデルを構築する。

In [195]:
kf = KFold(n_splits=5, shuffle=True, random_state=1)

In [191]:
predicted_cross_df = pd.DataFrame({'index':0, 'pred':0}, index=[1])
for train_index, val_index in kf.split(df_exp.index):
    
    #訓練データindexと評価データindexを使って、訓練データと評価データ＆説明変数と目的変数に分割
    X_train_cross = df_exp.iloc[train_index]
    y_train_cross = train_meta_df["views"].iloc[train_index]
    X_test_cross = df_exp.iloc[val_index]
    y_test_cross = train_meta_df["views"].iloc[val_index]
    
    #LightGBM高速化の為のデータセットに加工する
    lgb_cross_train = lgb.Dataset(X_train_cross, y_train_cross)
    lgb_cross_eval = lgb.Dataset(X_test_cross, y_test_cross)
    
    #LightGBMのモデル構築
    gbm_cross = lgb.train(params,
                lgb_cross_train,
                valid_sets=(lgb_cross_train, lgb_cross_eval),
                num_boost_round=1000,
                early_stopping_rounds=100,
                verbose_eval=50)
    
    #モデルに評価用説明変数を入れて、予測値を出力する
    predicted_cross = gbm_cross.predict(X_test_cross)
    
    #temp_df : 予測値と正答値を照合するために、予測値と元のindexを結合
    temp_df = pd.DataFrame({'index':X_test_cross.index, 'pred':predicted_cross})
    
    #predicted_df : 空のデータフレームにtemp_dfを結合→二周目以降のループでは、predicted_df（中身アリ）にtemp_dfが結合する
    predicted_cross_df = pd.concat([predicted_cross_df, temp_df], axis=0)

Training until validation scores don't improve for 100 rounds
[50]	training's rmse: 460.502	valid_1's rmse: 1210.87
[100]	training's rmse: 204.328	valid_1's rmse: 1219.37
Early stopping, best iteration is:
[16]	training's rmse: 803.84	valid_1's rmse: 1200.95
Training until validation scores don't improve for 100 rounds
[50]	training's rmse: 513.597	valid_1's rmse: 1022.82
[100]	training's rmse: 227.382	valid_1's rmse: 1032.62
Early stopping, best iteration is:
[17]	training's rmse: 847.234	valid_1's rmse: 1002.33
Training until validation scores don't improve for 100 rounds
[50]	training's rmse: 475.007	valid_1's rmse: 1219.55
[100]	training's rmse: 216.865	valid_1's rmse: 1234.67
Early stopping, best iteration is:
[21]	training's rmse: 761.067	valid_1's rmse: 1211.99
Training until validation scores don't improve for 100 rounds
[50]	training's rmse: 499.512	valid_1's rmse: 1006.78
[100]	training's rmse: 221.19	valid_1's rmse: 1018.07
Early stopping, best iteration is:
[15]	training's 

In [196]:
# 予測値を整理する

predicted_cross_df_cleaned = predicted_cross_df.sort_values("index").reset_index(drop = True).drop("index", axis = 1).drop(index = [0])
predicted_cross_df_cleaned

pred
1      415.271602
2      570.916439
3      379.329076
4      527.386286
5     1351.461440
...           ...
2996   643.639068
2997   310.298077
2998   595.021853
2999   512.366269
3000   330.297555

[3000 rows x 1 columns]

In [197]:
# 予測値に目的変数を加える

predicted_cross_df_cleaned["views"] = train_meta_df["views"].values
predicted_cross_df_cleaned

pred  views
1      415.271602    290
2      570.916439   1443
3      379.329076    329
4      527.386286     63
5     1351.461440     37
...           ...    ...
2996   643.639068     20
2997   310.298077    234
2998   595.021853     97
2999   512.366269     77
3000   330.297555     59

[3000 rows x 2 columns]

R2係数とRMSEを計算する。

In [198]:
R2_X_cross = r2_score(predicted_cross_df_cleaned["views"], predicted_cross_df_cleaned["pred"])
RMSE_X_cross = np.sqrt(mean_squared_error(predicted_cross_df_cleaned["views"], predicted_cross_df_cleaned["pred"]))

print("交差検証結果のR2係数 :" + str(R2_X_cross))
print("交差検証結果のRMSE係数 :" + str(RMSE_X_cross))

交差検証結果のR2係数 :0.13798920271572201
交差検証結果のRMSE係数 :1125.2322316060245


交差検証前の数字がR2約0.12, RMSE約1124だったので、R2係数は若干上がり、RMSEは若干下がった。

# Submission Fileを作成する

上記の交差検証を用いて、public dataから予測値を作成する。

In [207]:
# idの削除

df_test = df_test.drop("comp_id", axis = 1)
df_test

ad_blocked  embed    ratio  duration  language  partner  partner_active  \
0             0      1  1.33333      1675         2        0               0   
1             0      1  1.33333      1479         2        0               0   
2             0      1  1.33333      1505         2        0               0   
3             0      1  1.77778        50         2        1               1   
4             0      1  1.33333      1543         2        0               0   
..          ...    ...      ...       ...       ...      ...             ...   
981           0      1  1.77778        46         3        1               1   
982           0      1  1.33333        91         2        0               0   
983           0      1  1.77778        72         3        1               1   
984           0      1  1.33333        82         7        0               0   
985           0      1  1.34091        88         2        0               0   

     n_likes  n_tags  n_formats  ...   desc_41   desc_42   desc_43   desc_44  \
0          0      10          2  ... -1.111625  0.252028 -0.193523  0.346493   
1          0      10          2  ... -1.111625  0.252028 -0.193523  0.346493   
2          0      10          2  ... -1.111625  0.252028 -0.193523  0.346493   
3          1       6          5  ...  0.246156 -1.001942 -1.502183  1.105716   
4          0       6          2  ... -1.111625  0.252028 -0.193523  0.346493   
..       ...     ...        ...  ...       ...       ...       ...       ...   
981        0       2          4  ... -0.429883  2.516276  0.551086 -0.445840   
982        0       1          2  ...  2.279406  1.579310  6.187336 -0.259915   
983        0       2          4  ... -0.852350  2.475394  0.530240 -0.414277   
984        0       9          3  ... -1.924031 -0.009321  1.078347  0.409481   
985        0       4          3  ...  1.163483 -1.815711  0.958069  1.194779   

      desc_45   desc_46   desc_47   desc_48   desc_49   desc_50  
0   -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
1   -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
2   -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
3   -0.170937  0.181976  0.437707  0.058352 -2.209322  0.104916  
4   -0.077318  0.276308 -0.045740 -0.007150 -3.448430 -0.031262  
..        ...       ...       ...       ...       ...       ...  
981 -1.834950 -1.343484  1.518998  0.690137  2.181449  0.421095  
982  2.230830 -3.986988 -0.375218 -0.185296 -3.666707  0.629068  
983 -1.807584 -1.542420  1.573145  0.537607  2.239486  0.168694  
984  1.172790  0.548644 -0.209369 -0.329074  0.300092 -0.437185  
985  1.158204 -0.711968 -0.568707 -0.240866  0.391280 -0.164088  

[986 rows x 4112 columns]

In [215]:
# Public dataを、上記のモデルで交差検証する

predicted_submit_df = pd.DataFrame({'index':0, 'pred':0}, index=[1])
for train_index, val_index in kf.split(df_test.index):
    X_train_submit = df_test.iloc[val_index]
    
    predicted_submit = gbm_cross.predict(X_train_submit)

    temp_df = pd.DataFrame({'index':X_train_submit.index, 'pred':predicted_submit})

    predicted_submit_df = pd.concat([predicted_submit_df, temp_df], axis=0)

In [216]:
predicted_submit_cleaned = predicted_submit_df.sort_values("index").reset_index(drop = True).drop("index", axis = 1).drop(index = [0])
predicted_submit_cleaned

pred
1     353.137510
2     264.906811
3     352.063284
4    1000.959783
5     264.906811
..           ...
982   273.052722
983   651.680495
984   270.592750
985   924.114185
986   480.560864

[986 rows x 1 columns]

提出形式の確認

In [219]:
%cd C:\Users\Takeshi\OneDrive\Bitgrid\Dataset

C:\Users\Takeshi\OneDrive\Bitgrid\Dataset


In [223]:
solution_format = pd.read_csv("solution_format.csv")
solution_format

comp_id  views
0       3001    879
1       3002   6649
2       3003   2249
3       3004   6586
4       3005    973
..       ...    ...
981     3982   9539
982     3983   2025
983     3984   4324
984     3985    925
985     3986   2821

[986 rows x 2 columns]

In [230]:
#提出ファイルの作成

submission = solution_format.copy()
submission["views"] =  predicted_submit_cleaned["pred"].values
submission

comp_id        views
0       3001   353.137510
1       3002   264.906811
2       3003   352.063284
3       3004  1000.959783
4       3005   264.906811
..       ...          ...
981     3982   273.052722
982     3983   651.680495
983     3984   270.592750
984     3985   924.114185
985     3986   480.560864

[986 rows x 2 columns]

In [235]:
submission.to_csv("submission_7.2.2021.csv", index = False)

# 評価方法の計算式で計算する

評価方法であるe^(-RMSE/MAX(observed_values))の計算結果を出す。

In [236]:
predicted_cross_df_cleaned["pred"].max()

2999.422611414366

In [237]:
math.exp(-RMSE_X_cross/predicted_cross_df_cleaned["pred"].max())

0.6871864591382114